File: Async/AsyncAndroid/app/src/main/java/com/bpxamarin/asyncandroid/MainActivity.java
package com.bpxamarin.asyncandroid;

import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

public class MainActivity extends Activity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button btnAsyncTask = findViewById(R.id.btn_async_task);
        btnAsyncTask.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                executeAsyncTask();
            }
        });
    }

    private void executeAsyncTask(){
        Toast.makeText(this, "Started async task", Toast.LENGTH_SHORT).show();
        new WaitTask().execute(this);
    }

    private static class WaitTask extends AsyncTask<Context, Void, Context> {
        @Override
        protected Context doInBackground(Context... params) {
            Context context = params[0];
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            return context;
        }

        @Override
        protected void onPostExecute(Context context) {
            Toast.makeText(context, "Finished async task", Toast.LENGTH_SHORT).show();
        }
    }
}

File: Async/AsyncXamarin/App.Droid/MainActivity.cs
using System.Threading.Tasks;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Views;
using Android.Widget;

namespace App.Droid
{
    [Activity(MainLauncher = true)]
    public class MainActivity : Activity, View.IOnClickListener
    {
        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.activity_main);

            var btnAsyncTask = FindViewById(Resource.Id.btn_async_task);
            var btnAsyncMethod = FindViewById(Resource.Id.btn_async_method);

            btnAsyncTask.SetOnClickListener(this);
            btnAsyncMethod.SetOnClickListener(this);
        }

        private async void ExecuteAsyncMethod()
        {
            Toast.MakeText(this, "Started async method", ToastLength.Short).Show();
            await Task.Run(() =>
            {
                System.Threading.Thread.Sleep(3000);
            });
            Toast.MakeText(this, "Finished async method", ToastLength.Short).Show();
        }

        private void ExecuteAsyncTask()
        {
            Toast.MakeText(this, "Started async task", ToastLength.Short).Show();
            new WaitTask().Execute(this);
        }

        private class WaitTask : AsyncTask {
            protected override Java.Lang.Object DoInBackground(params Java.Lang.Object[] @params)
            {
                var context = @params[0];
                Java.Lang.Thread.Sleep(3000);
                return context;
            }

            protected override void OnPostExecute(Java.Lang.Object result)
            {
                Toast.MakeText((Context)result, "Finished async task", ToastLength.Short).Show();
            }
        }

        public void OnClick(View v)
        {
            switch (v.Id)
            {
                case Resource.Id.btn_async_task:
                    ExecuteAsyncTask();
                    break;
                case Resource.Id.btn_async_method:
                    ExecuteAsyncMethod();
                    break;
            }
        }
    }
}

